### ISSUE TITLE
New adverbs `at()` and `where()`

### ISSUE THREAD

**Issue by lionel-** [hadley/purrr#56]
One fundamental limitation of replacement functions such as `[<-` is that they cannot replace an element with a larger one.  i.e., it's illegal to do this:
```{r}
mtcars[3] <- data.frame(a = mtcars[[3]], b = mtcars[[3]])
```

This is because it would be impossible to know what goes where in case the replacement region is not contiguous. But it makes it hard to work with functions that take a data frame column and may return a larger data frame, like `disjoin()` (see #28 for some background on this tool). We really need a way to apply this kind of function on subsets of columns.

So I wrote these two adverbs inspired by `fapply()`:
```{r}
at <- function(.x, .where, ..f, ...) {
  if (is.character(.where)) {
    if (is.null(names(.x))) {
      stop(".x is not named", call. = FALSE)
    }
    .where <- match(.where, names(.x))
  }
  if (anyNA(.where)) {
    stop("some indexes are missing", call. =FALSE)
  }

  out <- vector("list", length(.x))
  for (i in seq_along(.x)) {
    res <-
      if (i %in% .where) {
        ..f(.x[i], ...)
      } else {
        .x[i]
      }
    stopifnot(is.list(res))
    out[[i]] <- res
  }

  flatten(out) %>% purrr:::output_hook(.x)
}

where <- function(.x, .p, ..f, ...) {
  sel <- probe(.x, .p) %>% which()
  .x %>% at(sel, ..f, ...)
}
```

For example,
```{r}
diamonds %>% at("cut", disjoin, sep = " / ") %>% str()
## Classes ‘tbl_df’, ‘tbl’ and 'data.frame':	53940 obs. of  14 variables:
##  $ carat          : num  0.23 0.21 0.23 0.29 0.31 0.24 0.24 0.26 0.22 0.23 ...
##  $ cut / Fair     : num  0 0 0 0 0 0 0 0 1 0 ...
##  $ cut / Good     : num  0 0 1 0 1 0 0 0 0 0 ...
##  $ cut / Very Good: num  0 0 0 0 0 1 1 1 0 1 ...
##  $ cut / Premium  : num  0 1 0 1 0 0 0 0 0 0 ...
##  $ cut / Ideal    : num  1 0 0 0 0 0 0 0 0 0 ...
##  $ color          : Ord.factor w/ 7 levels "D"<"E"<"F"<"G"<..: 2 2 2 6 7 7 6 5 2 5 ...
##  $ clarity        : Ord.factor w/ 8 levels "I1"<"SI2"<"SI1"<..: 2 3 5 4 2 6 7 3 4 5 ...
##  $ depth          : num  61.5 59.8 56.9 62.4 63.3 62.8 62.3 61.9 65.1 59.4 ...
##  $ table          : num  55 61 65 58 58 57 57 55 61 61 ...
##  $ price          : int  326 326 327 334 335 336 336 337 337 338 ...
##  $ x              : num  3.95 3.89 4.05 4.2 4.34 3.94 3.95 4.07 3.87 4 ...
##  $ y              : num  3.98 3.84 4.07 4.23 4.35 3.96 3.98 4.11 3.78 4.05 ...
##  $ z              : num  2.43 2.31 2.31 2.63 2.75 2.48 2.47 2.53 2.49 2.39 ...
```
transforms "cut" into a disjunctive multi-columns form and
```{r}
diamonds %>% where(is.factor, disjoin)
```
would transform all factors.

Then, `map_if()` could be implemented in terms of `where()`, and we could also have a `map_at()` version:
```{r}
map_if <- function(.x, .p, .f, ...) {
  .f <- purrr:::as_function(.f)
  .x %>% where(.p, map, .f, ...)
}

map_at <- function(.x, .where, .f, ...) {
  .f <- purrr:::as_function(.f)
  .x %>% at(.where, map, .f, ...)
}
```

This would not change the functionality of `map_if()` in any way, but make the internal code a bit dryer.

I can document these tools and send a PR.

**Comment by hadley** [comment]
I like it!

**Comment by lionel-** [comment]
Great :-)

now that I think of it, the purrr version of `fapply()` could be implemented in terms of `at()`.
But I really have trouble finding a good name for this one. I don't like `fmap()` too much, and `remap()` is also not that good.

**Comment by lionel-** [comment]
Maybe add an argument in `map()`?

```r
mtcars %>% map(disjoin, .preserve = TRUE)
mtcars %>% map(disjoin, .as_list = TRUE)
```

Or `self_map()`, `map_list()`, `outer_map()`?

**Comment by lionel-** [comment]
`at_each`? Would go well by analogy with `at()`. The downside is that we loose the semantic link with `map()`, but I think it's a good option.

```r
mtcars %>% at_each(disjoin)
```

 Actually, maybe it's not a bad thing to loose the semantic connection with `map()`, maybe it will be less confusing for people new to purrr or functional programming.

**Comment by lionel-** [comment]
On second thought, these are not adverbs. At least, not more than `map()` is. They are mapping functionals in themselves, with the particularity of being conditional and list-preserving.

Would it make sense to have them truly act as adverbs by not mapping over the selected regions? That is, all contiguous parts of the selected region would be fed in turn to `.f`. For example,
```r
mtcars %>% at(c(2, 4:8), disjoin)
```
would be equivalent to:
```r
c(
  mtcars[1],
  disjoin(mtcars[2]),
  mtcars[3],
  disjoin(mtcars[4:8]),
  mtcars[9:11]
)
```
Then they can be combined with actual mapping functionals.

And then it would make more sense to have purrr's preserving-map called something like `map_each()` rather than `at_each()`:
```r
mtcars %>% at(c(2:3, 6:9), map_each, disjoin)
diamonds %>% where(is.factor, map_each, disjoin)
```

I don't know if there would be many use-case for using `at()` and `where()` without mapping, though. But this way we create a more coherent set of complementary and composable list-preserving tools.

**Comment by lionel-** [comment]
Closed with #57
